x-pack/plugins/lens/public/app_plugin/app.test.tsx

```diff
@@ -495,190 +513,226 @@ describe('Lens App', () => {
             isSaveable: true,
           })
         );
-        instance.update();
+        component.update();
+        expect(getButton(component).disableButton).toEqual(false);
+
+        await act(async () => {
+          const topNavMenuConfig = component.find(TopNavMenu).prop('config');
+          expect(topNavMenuConfig).not.toContainEqual(
+            expect.objectContaining(navMenuItems.expectedSaveAndReturnButton)
+          );
+          expect(topNavMenuConfig).not.toContainEqual(
+            expect.objectContaining(navMenuItems.expectedSaveAsButton)
+          );
+          expect(topNavMenuConfig).toContainEqual(
+            expect.objectContaining(navMenuItems.expectedSaveButton)
+          );
+        });
+      });
+
+      it('Shows Save and Return and Save As buttons in create by value mode', async () => {
+        const props = makeDefaultProps();
+        const services = makeDefaultServices();
+        services.dashboardFeatureFlag = { allowByValueEmbeddables: true };
+        props.incomingState = {
+          originatingApp: 'ultraDashboard',
+          valueInput: {
+            id: 'whatchaGonnaDoWith',
+            attributes: {
+              title:
+                'whatcha gonna do with all these references? All these references in your value Input',
+              references: [] as SavedObjectReference[],
+            },
+          } as LensByValueInput,
+        };
+
+        const { component } = mountWith({ props, services });
 
-        expect(getButton(instance).disableButton).toEqual(false);
+        await act(async () => {
+          const topNavMenuConfig = component.find(TopNavMenu).prop('config');
+          expect(topNavMenuConfig).toContainEqual(
+            expect.objectContaining(navMenuItems.expectedSaveAndReturnButton)
+          );
+          expect(topNavMenuConfig).toContainEqual(
+            expect.objectContaining(navMenuItems.expectedSaveAsButton)
+          );
+          expect(topNavMenuConfig).not.toContainEqual(
+            expect.objectContaining(navMenuItems.expectedSaveButton)
+          );
+        });
       });
 
       it('saves new docs', async () => {
-        const { args, instance: inst } = await save({
-          initialDocId: undefined,
+        const { props, services, component } = await save({
+          initialSavedObjectId: undefined,
           newCopyOnSave: false,
           newTitle: 'hello there',
         });
-
-        expect(args.docStorage.save).toHaveBeenCalledWith(
+        expect(services.attributeService.wrapAttributes).toHaveBeenCalledWith(
           expect.objectContaining({
-            id: undefined,
+            savedObjectId: undefined,
             title: 'hello there',
-          })
+          }),
+          true,
+          undefined
         );
-
-        expect(args.redirectTo).toHaveBeenCalledWith('aaa', undefined, true);
-
-        inst.setProps({ docId: 'aaa' });
-
-        expect(args.docStorage.load).not.toHaveBeenCalled();
```

Comment:
I have removed this expect due to a change in behaviour that would make it fail. 

**but why???**
In older versions of lens, the state was updated on save with the new persisted / lastKnown doc even in cases when it would immediately redirect to the new `docId`

Here it is in master....

https://github.com/elastic/kibana/blob/17fec25e3c281b4f66e5973efbaef0368b9152c5/x-pack/plugins/lens/public/app_plugin/app.tsx#L376-L387

From my research it looks like it's been this way since the original save modal was added to lens. I am not sure when it started causing problems, but it looks like it might have contributed to why exhaustive-deps was disabled on the document load `useEffect` in master. 

**The change**
When I changed `runSave` to only update the `persistedDoc` when it is _not_ redirecting, I could fulfill exhaustive-deps and avoid a double-loading issue.

During the jest tests, the actual 'redirect' is not triggered, which means that the `persistedDoc` doesn't get updated, which means that updating the `savedObjectId` in this test would trigger another load of the document. Thankfully, this part of the test seems a bit redundant, since the behaviour is already covered by the test
`does not load documents on sequential renders unless the id changes `

Please let me know if I've missed or misrepresented anything